linker/linker_soinfo.h

```diff
@@ -338,6 +340,40 @@ struct soinfo {
   uintptr_t handle_;
 
   friend soinfo* get_libdl_info(const char* linker_path, const link_map& linker_map);
+
+ public:
+  struct SegmentInfo {
+    ElfW(Addr) phdr_addr, real_addr;
+    ElfW(Word) real_size, page_size;
+    size_t index;
+  };
+
+  typedef std::vector<SegmentInfo> seginfo_list_t;
+
+ public:
+  seginfo_list_t rand_addr_segments;
+
+  void sort_rand_addr_segments();
+
+  // Find and return a pagerando randomly mapped segment containing the given
+  // virtual address
+  SegmentInfo* find_rand_segment_info(ElfW(Addr) addr) {
```

Comment:
Is this really simpler than the loop? I don't really think it is.

    SegmentInfo* find_rand_segment(ElfW(Addr) mem_vaddr) {
      auto I = std::find_if(
        rand_addr_segments.begin(), rand_addr_segments.end(),
        [mem_vaddr](const SegmentInfo& seg_info) {
          ElfW(Addr) bin_start = seg_info.real_addr + PAGE_OFFSET(seg_info.phdr_addr);
          return (mem_vaddr >= bin_start &&
                  mem_vaddr < (bin_start + seg_info.real_size));
        });
      return I == rand_addr_segments.end() ? nullptr : &(*I);
    }